src/continuous_agg.c

```diff
@@ -213,6 +214,203 @@ continuous_agg_init(ContinuousAgg *cagg, const Form_continuous_agg fd)
 	Assert(OidIsValid(cagg->partition_type));
 }
 
+TSDLLEXPORT CaggsInfo
+ts_continuous_agg_get_all_caggs_info(int32 raw_hypertable_id)
+{
+	CaggsInfo all_caggs_info;
+
+	List *caggs = ts_continuous_aggs_find_by_raw_table_id(raw_hypertable_id);
+	ListCell *lc;
+	int64 *bucket_width;
+
+	all_caggs_info.bucket_widths = NIL;
+	all_caggs_info.max_bucket_widths = NIL;
+	all_caggs_info.mat_hypertable_ids = NIL;
+	Assert(list_length(caggs) > 0);
+	ContinuousAgg *last_cagg = llast(caggs);
+	all_caggs_info.last_mat_hypertable_id = last_cagg->data.mat_hypertable_id;
+
+	foreach (lc, caggs)
+	{
+		ContinuousAgg *cagg = lfirst(lc);
+
+		bucket_width = palloc(sizeof(*bucket_width));
```

Comment:
This allocates memory proportional to the number of caggs. While this might not be a problem for many use cases in practice, I would worry about this for applications that try to scale to many caggs (1000s). 

I would consider trying to structure the code so that you can actually nest the functionality you need within the loop and use a memory context that you can reset in each iteration of the loop instead.

Reply:
This is a direct consequence of the current design. The existing code in master allocates memory proportionally as well via `get_cagg_ids(int32 hyper_id)`. In the new code, instead of having a single integer list we have 3 integer lists.

The design decisions were the following:
- we don't have catalog information in the Data-Nodes
- we don't move back and forth between Access-Node and Data-Nodes, but we offload the processing to be done locally at the Data-Node.

As this is the first (distributed) transaction that holds a heavy exclusive lock in the Acess-Node we attempt to finish invalidation processing as quickly as possible, in the Data-Nodes by processing hypertable log invalidations in one pass for all CAGGs of the underlying distributed hypertable.

Nesting the iterations as you propose would indeed solve the memory problem, but this was done intentionally for the above reasons.